pkg/dns/dnscache.go

```diff
@@ -0,0 +1,258 @@
+package dns
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"net"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/google/nftables"
+	dnsgo "github.com/miekg/dns"
+)
+
+const (
+	tableName = "firewall"
+)
+
+type ipEntry struct {
+	ips            []net.IP
+	expirationTime time.Time
+	setName        string
+}
+
+func (e *ipEntry) getNewAndDeletedIPs(ips []net.IP) (newIPs, deletedIPs []nftables.SetElement) {
+	currentIps := make(map[string]bool, len(e.ips))
+	for _, ip := range e.ips {
+		currentIps[ip.String()] = false
+	}
+
+	for _, ip := range ips {
+		s := ip.String()
+		if _, ok := currentIps[s]; ok {
+			currentIps[s] = true
+		} else {
+			newIPs = append(newIPs, nftables.SetElement{Key: ip})
+		}
+	}
+
+	for ip, exists := range currentIps {
+		if !exists {
+			deletedIPs = append(deletedIPs, nftables.SetElement{Key: net.ParseIP(ip)})
+		}
+	}
+
+	return
+}
+
+type cacheEntry struct {
+	ipv4 *ipEntry
+	ipv6 *ipEntry
+}
+
+type DNSCache struct {
+	fqdnToEntry map[string]cacheEntry
+	setNames    map[string]struct{}
+}
+
+// TODO GetIPs returns list of IPs for FQDN
+func (c *DNSCache) GetIPs(fqdn string) []string {
+	return nil
```

Comment:
For plain FQDNs without regex we should actively lookup the FQDN (if it's not in the cache) - instead of relying on data captured with the transparent DNS proxy.